### PR TITLE
Don't expand groups for trailing end-of-line comments

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

This is a PR where we need to align on desired behavior. Right now, we expand a group when writing a trailing end-of-line comment, so e.g., given:

```python
{
    a: a  # a
    for c in e  # for # c 
}
```

We format this expression as:

```python
{
    a: a  # a
    for c in e  # for # c
}
```

Black, meanwhile, doesn't let trailing end-of-line comments expand, so formats as:

```python
{a: a for c in e}  # a  # for # c
```

Looking through the snapshot diff, this seems to strictly improve Black compatibility (and our current behavior seems like a significant deviation). Even the changes in our own fixtures now better resemble Black as verified in the playground.

I don't know that I have a super strong opinion on whether to use Black's behavior or our own, but I lead towards following Black's behavior by default.

## Test Plan

Weird mix on the similarity scores -- some went down a little, some went up a little.

Before:

- `zulip`: 0.99702
- `django`: 0.99784
- `warehouse`: 0.99585
- `build`: 0.75623
- `transformers`: 0.99469
- `cpython`: 0.75988
- `typeshed`: 0.74853

After:

- `zulip`: 0.99696
- `django`: 0.99785
- `warehouse`: 0.99569
- `build`: 0.75623
- `transformers`: 0.99481
- `cpython`: 0.75987
- `typeshed`: 0.74842

### CODE REVIEW

File: crates/ruff_python_formatter/tests/snapshots/format@expression__compare.py.snap [hunk 1]
```diff
@@ -138,10 +138,7 @@ a not in b
     == b
 )
 
-(
-    a  # comment
-    == b
-)
+(a == b)  # comment
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/tests/snapshots/format@expression__compare.py.snap:141]: (This now matches Black.)

File: crates/ruff_python_formatter/tests/snapshots/format@expression__dict.py.snap [hunk 2]
```diff
@@ -78,14 +78,9 @@ x={  # dangling end of line comment
 
 {**d}
 
-{
-    **a,  #  leading
-    **b,  # middle  # trailing
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/tests/snapshots/format@expression__dict.py.snap:83]: I think we have a separate bug related to trailing commas on `**` arguments.

File: crates/ruff_python_formatter/tests/snapshots/format@expression__dict_comp.py.snap [hunk 3]
```diff
@@ -128,10 +128,7 @@ aaaaaaaaaaaaaaaaaaaaa = {
     ]
 }
 
-{
-    a: a  # a
-    for c in e  # for  # c  # in  # e
-}
+{a: a for c in e}  # a  # for  # c  # in  # e
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/tests/snapshots/format@expression__dict_comp.py.snap:131]: This now matches Black.

File: crates/ruff_python_formatter/tests/snapshots/black_compatibility@simple_cases__comments2.py.snap [hunk 4]
```diff
@@ -382,11 +340,7 @@ short
         # yup
         arg3=True,
     )
-    lcomp = [
-        element  # yup
-        for element in collection  # yup
-        if element is not None  # right
-    ]
+    lcomp = [element for element in collection if element is not None]  # yup  # yup  # right
```
Comment (charliermarsh) [inline crates/ruff_python_formatter/tests/snapshots/black_compatibility@simple_cases__comments2.py.snap:343]: This is much closer to Black but still wrong due to the line suffixes not being included when determining the line length:

```python
lcomp = [
    element for element in collection if element is not None  # yup  # yup  # right
]
```

**Comment by MichaReiser** [comment]
Interesting to see this change in action. 

Our formatter goal is to have (close) Black parity for already formatted code, but we intentionally allow deviation for unformatted code (using Black and Ruff in the same project isn't a use case we want to support). I don't expect this change to help with our goal because Black should already have moved all comments to the end of the line. That's why I'm surprised that it affects the similarity index at all (the overall change seems neutral). We'll have to manually investigate the changes before moving forward. I suspect that these are accidental changes where we have a different comment placement and the lines now happen to be short enough so that we place the comment at the same position as Black.

The main question to me is whether we believe that flattening as many elements as possible and collapsing comments at the end is the better formatting. In my view this isn't the case because comments are moved away from where I intentionally placed them:

### Pragma comments

```python
def test(
	a, #  type-ignore
	b
):
	pass
```

This now gets formatted as 

```python
def test(a, b): #  type-ignore
	pass
```

which not only suppresses typing errors for `a`, but also `b`. I now have to go back, expand the parameters and add a trailing comma after `b` if I want to keep the suppression specific to `a`, which is what I intended. This is a lot of work.

The workaround with adding trailing commas also only works in positions where they are supported and if you have more than one argument. For example, the following doesn't work

```python
(
    call(has_type_error) # type-ignore
    and another_call()
)
```

gets formatted as 

```python
(call(has_type_error) and another_call())  # type-ignore
```

which suppresses typing errors in `another_call` too. I doubt anyone would bother enough (until they have a bug because of the missed type checker error) to undo Black's change and add suppression comments:

```python
# This doesn't work, probably because we have collapsed comments
(
    call(has_type_error) # fmt: skip # type-ignore 
    and another_call()
)

(
    # fmt: off
    call(has_type_error)  # type-ignore
    # fmt: on
    and another_call()
)
```

We won't be able to provide this escape hatch because Ruff doesn't support suppression comments on the expression level (you would need to suppress the whole statement). And suppression comments are a net negative in my view. 


### Comment context

This is related to pragma comments. Moving comments too far means that the comments loos their context. 

```python
call(
	[], #  TODO fill in values
	b,
	[]
)
```

gets formatted to

```python
call([], b, [])  #  TODO fill in values
```

It is now unclear if I have to fill in the values in the first or last list. Again, I can work around this by adding a trailing comma but that's a lot of work to make the formatter respect my comment placement.

Similar

```python
call(
	[], # Empty because of X
	c,
	[], # Empty because of Y
)
```

gets formatted as 

```python
call([], c, [])  # Empty because of X  # Empty because of Y
```

Again, we have the context problem, but I also dislike collapsed comments `# comment A # comment B` looks strange to me. It also seems to bring that it breaks pragma comments, as we've seen with Black's `fmt: skip` pragma comment.

**Review by konstin — APPROVED** [review]
i'm not really opinionated either way, both styles have their advantages and disadvantages to me

**Comment by charliermarsh** [comment]
We're not making this change, for now at least.

**Pull request closed without merging**